python/ray/data/_internal/execution/operators/map_operator_tasks_impl.py

```diff
@@ -0,0 +1,174 @@
+                self._block_bundle = bundle
+            else:
+                self._create_task(bundle)
+        else:
+            self._block_bundle = _merge_ref_bundles(self._block_bundle, bundle)
+
+    def inputs_done(self, input_index: int) -> None:
+        self._input_deps_done += 1
+        assert self._input_deps_done <= len(self._op._input_dependencies)
+        if (
+            self._input_deps_done == len(self._op._input_dependencies)
+            and self._block_bundle
+        ):
+            self._create_task(self._block_bundle)
+            self._block_bundle = None
+
+    def work_completed(self, ref: ObjectRef[ObjectRefGenerator]) -> None:
+        task = self._tasks.pop(ref)
+        all_refs = list(ray.get(ref))
+        block_refs = all_refs[:-1]
+        block_metas = ray.get(all_refs[-1])
+        del ref
+        assert len(block_metas) == len(block_refs), (block_refs, block_metas)
+        for ref in block_refs:
+            trace_allocation(ref, "map_operator_work_completed")
+        task.output = RefBundle(list(zip(block_refs, block_metas)), owns_blocks=True)
+        allocated = task.output.size_bytes()
+        self._obj_store_mem_alloc += allocated
+        self._obj_store_mem_cur += allocated
+        # TODO(ekl) this isn't strictly correct if multiple operators depend on this
+        # bundle, but it doesn't happen in linear dags for now.
+        freed = task.inputs.destroy_if_owned()
+        if freed:
+            self._obj_store_mem_freed += freed
+            self._obj_store_mem_cur -= freed
+        if self._obj_store_mem_cur > self._obj_store_mem_peak:
+            self._obj_store_mem_peak = self._obj_store_mem_cur
+
+    def has_next(self) -> bool:
+        i = self._next_output_index
+        return (
+            i in self._tasks_by_output_order
+            and self._tasks_by_output_order[i].output is not None
+        )
+
+    def get_next(self) -> RefBundle:
+        i = self._next_output_index
+        self._next_output_index += 1
+        bundle = self._tasks_by_output_order.pop(i).output
+        self._obj_store_mem_cur -= bundle.size_bytes()
+        return bundle
+
+    def get_work_refs(self) -> List[ray.ObjectRef]:
+        return list(self._tasks)
+
+    def shutdown(self) -> None:
+        # Cancel all active tasks.
+        for task in self._tasks:
+            ray.cancel(task)
+        # Wait until all tasks have failed or been cancelled.
+        for task in self._tasks:
+            try:
+                ray.get(task)
+            except ray.exceptions.RayError:
+                # Cancellation either succeeded, or the task had already failed with
+                # a different error, or cancellation failed. In all cases, we
+                # swallow the exception.
+                pass
+
+    def _create_task(self, bundle: RefBundle) -> None:
+        input_blocks = []
+        for block, _ in bundle.blocks:
+            input_blocks.append(block)
```

Comment:
Previously we have a nice warning for users in case blocks bundling reduces parallelism too much ([code](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/compute.py#L519-L527)). Is it possible to have a simliar warning here?

Reply:
This is a little trickier in the streaming execution model, maybe we can add it back later as needed.

Reply:
I think that's pretty easy to do with either executor: instead of looking at whether there is a 10x reduction in parallelism globally after block bundling, we keep a counter for the size of the of the **current** bundle, and if we've merged more than 10 bundles into a single bundle, we log a warning (if we haven't already logged a warning, i.e. log-once). Or we could keep a running average and log at the end of execution, although I think that the eager logging would be better.

That should work for both the bulk executor and the streaming executor, right?

Reply:
Let me add a TODO to do this later.